[#20] 친구 목록 불러오기 기능 추가

msugo1 opened the pull request:
## 📝 페이징 적용
-  친구 수가 엄청나게 많은 사용자가 있다고 가정하면 `모든 친구를 한 번에 불러오는 것은 비효율적`이기도 하고, 데이터를 불러올 때 `로딩에도 영향`을 미친다고 생각했고 DB에 조건에 맞는 데이터가 많을 수록 페이징은 필수라고 생각해서 적용했습니다.

## 📖 Model 관련

### 📑  FriedListRequest

- 친구목록 불러오기에 필요한 값들을 담아서 DB로 전달할 객체를 생성하기 위한 클래스

- pageNumber 가 `음수로 들어와도 기본값을 1로 설정`해서 맨 첫 페이지가 보이도록 설정

### 📑 FriendList

- DB로 부터 가져온 `결과 값을 담을 객체`를 생성할 클래스(친구 ID와 친구 상태를 인스턴스 변수로 가지고 있음)

- immutable 객체 상태로 직렬화가 가능하게 만드는 Lombok `@Value`  어노테이션 활용

## 📝 상수 관련

### 📑 Pagination

- 페이징 시 항목 별로 최대 페이지 값을 담아두기 위한 클래스 생성

## 📝 컨트롤러 계층 관련

### 📑 FriendController: getFriendList 메소드 추가

- 현재 사용자의 친구 목록을 불러오기 위한 요청을 매핑할 메소드

- 첫 화면을 표시하기 위한 pageNumber `기본값` 설정

- 받아온 친구목록을 `ResponseEntity`를 활용해 본문에 담아서 화면에 반환

## 📝 서비스 계층 관련

### 📑 FriendService: getFriendList 메소드 추가

- 들어온 요청을 바탕으로 `FriendListRequest` 객체를 생성해 Mapper에 전달

- 반환된 값은 Controller에 전달

## 📝 매퍼 계층 관련

### 📑 FriendMapper: getFriendList 메소드 추가

- DB에서 친구목록을 불러오기 위한 메소드

### 📑 FriemdMapper.xml

- 위의 로직을 처리 할 쿼리문 추가

Commit:
[#20] 친구 목록 불러오기 기능 추가

- 페이징 적용
Why? 친구 수가 엄청나게 많은 사용자가 있다고 가정하면 모든 친구를 불러오는
것은 비효율적이고, 데이터를 불러올 때 데이터가 불필요하게 많아져
로딩에도 영향을 미칠 것

Model 관련

* FriedListRequest

- 친구목록 불러오기에 필요한 값들을 담아서 DB로 전달할 객체를 생성하기 위한
  클래스

- pageNumber 가 `음수로 들어와도 기본값을 1로 설정`해서 맨 첫 페이지가 보이도록 설정

* FriendList

- DB로 부터 가져온 `결과 값을 담을 객체`를 생성할 클래스(친구 ID와 친구
  상태를 인스턴스 변수로 가지고 있음)

- immutable 객체 상태로 직렬화가 가능하게 만들고자 Lombok `@Value`
  어노테이션 활용

상수 관련

* Pagination

- 페이징 시 항목 별로 최대 페이지 값을 담아두기 위한 클래스 생성

컨트롤러 계층 관련

* FriendController: getFriendList 메소드 추가

- 현재 사용자의 친구 목록을 불러오기 위한 요청을 매핑할 메소드

- 첫 화면을 표시하기 위한 pageNumber 기본값 설정

- 받아온 친구목록을 `ResponseEntity`를 활용해 본문에 담아서 화면에 반환

서비스 계층 관련

* FriendService: getFriendList 메소드 추가

- 들어온 요청을 바탕으로 `FriendListRequest` 객체를 생성해 Mapper에 전달

- 반환된 값은 Controller에 전달

매퍼 계층 관련

* FriendMapper: getFriendList 메소드 추가

- DB에서 친구목록을 불러오기 위한 메소드

* FriemdMapper.xml

- 위의 로직을 처리 할 쿼리문 추가

Commit:
[#20] 주석 추가 및 오탈자 수정

Commit:
[#21] 받은 친구 요청 목록 불러오기 기능 구현

컨트롤러 계층 관련

* FriendController: getFriendRequestList 메소드 추가

- 현재 사용자의 친구 목록을 불러오기 위한 요청을 매핑할 메소드

- 첫 화면을 표시하기 위한 pageNumber `기본값` 설정

- 받아온 친구목록을 `ResponseEntity`를 활용해 본문에 담아서 화면에 반환

서비스 계층 관련

* FriendService: getFriendRequestList 메소드 추가

- 들어온 요청을 바탕으로 `FriendListRequest` 객체를 생성해 Mapper에 전달

- 반환된 값은 Controller에 전달

f-lab-dev commented on src/main/java/me/soo/helloworld/controller/FriendController.java:
```diff
@@ -44,4 +48,12 @@ public void rejectFriendRequest(@CurrentUser String userId, @PathVariable String
+    @LoginRequired
+    @GetMapping
+    public ResponseEntity<List<FriendList>> getFriendList(@CurrentUser String userId,
+                                                          @RequestParam(defaultValue = "1") Integer pageNumber) {
```
`defaultValue` 좋네요~

f-lab-dev commented on src/main/java/me/soo/helloworld/model/friend/FriendList.java:
```diff
@@ -0,0 +1,12 @@
+import lombok.Value;
+import me.soo.helloworld.enumeration.FriendStatus;
+
+@Value
```
혹시나 해서 여쭤보자면 이 어노테이션은 어떤 의미일까요?

msugo1 replied:
`@Data`의 `immutability` 버전이라고 부를 수 있습니다.
1. 해당 어노테이션을 붙일 경우 필드에 있는 변수에 기본적으로  `private` 접근제어자와 `final` 예약어가 붙습니다.
2. class 자체도 `final`이 되어 상속이 불가능해집니다.
3. `@Data`의 immutability 버전이기 때문에 `@Data`와는 다르게 `Setter`는 생성하지 않습니다.

f-lab-dev commented on src/main/java/me/soo/helloworld/util/Pagination.java:
```diff
@@ -0,0 +1,6 @@
+
+public class Pagination {
+
+    public static final int MAX_PER_PAGE_FRIEND = 30;
```
추후 이 값이 바뀔 여지가 있다면 외부에서 주입해주도록 하는것도 좋은 방법입니다~

msugo1 replied:
원래 구상은 요청별로 제한 수를 다르게 설정해주는 것이었는데 그래도 외부에서 주입받는 것이 좀 더 유연한 방법일까요?

f-lab-dev replied:
이거는 선택사항입니다~ 기획에 따라 달라질 수 있습니다

Commit:
[#21] 피드백 반영

FriendController

- URL 불필요한 부분 제거
(/my-friends/friend-requests/from/other-users ->
/my-friends/friend-requests)

- 불필요한 ResponseEntity 리턴제거

Commit:
[#20] 피드백 반영

FriendController

* getFriendList 메소드

- 불필요한 ResponseEntity 리턴 제거

msugo1 committed:
Merge pull request #47 from f-lab-edu/feature/21

[#21] 받은 친구 요청 목록 불러오기 기능 구현

Commit:
[#20] 피드백 반영

* 생성자와 @Value의 조합을 가지고 외부에서 pagination에 대한 max Page
  값을 주입 받을 수 있도록 설정

Commit:
[#20] git push 충돌 해결

- Merge branch 'feature/20' of github.com:f-lab-edu/Hello-World into feature/20

msugo1 committed:
Merge branch 'develop' into feature/20
